rasterio/_io.pyx

```diff
@@ -746,21 +788,23 @@ cdef class DatasetReaderBase(DatasetBase):
             'boundless': boundless,
             'resampling': resampling}
 
-        # GDAL found dataset-wide alpha band or mask
-        # All band masks are equal so we can return the first
         if MaskFlags.per_dataset in self.mask_flag_enums[0]:
             return self.read_masks(1, **kwargs)
 
-        # use Alpha mask if available and looks like RGB, even if nodata is shadowing
         elif self.count == 4 and self.colorinterp[0] == ColorInterp.red:
             return self.read_masks(4, **kwargs)
 
-        # Or use the binary OR intersection of all GDALGetMaskBands
-        else:
-            mask = self.read_masks(1, **kwargs)
-            for i in range(1, self.count):
-                mask = mask | self.read_masks(i, **kwargs)
```

Comment:
The problem is with `range(1, self.count)`. We've already read the mask for `i = 1` and so we're repeating some work. And then we skip the final band of the dataset, which has index `count`. Python's `range` stops before the second index.